arxiv/base/urls.py

```diff
@@ -0,0 +1,73 @@
+"""Helpers for working with URLs in arXiv Flask applications."""
+
+from urllib.parse import parse_qs
+from werkzeug.urls import url_encode, url_parse, url_unparse, url_encode
+from flask import current_app
+from arxiv.base import config
+from arxiv.base.exceptions import ConfigurationError
+
+
+def config_url(target: str, path_params: dict = {}, params: dict = {}) -> str:
+    """
+    Generate a URL from this app's configuration.
+
+    This will load the value of the configuration parameter
+    `ARXIV_{target}_URL` from either the current application configuration
+    (preferred), or the base configuration defined in this package (see
+    :mod:`arxiv.base.config`).
+
+    Note: this function relies on the `flask.current_app` proxy object, which
+    means that it can only be used in an application or request context.
+
+    Parameters
+    ----------
+    target : str
+        Name of the endpoint, defined as `ARXIV_{target}_URL` in the
+        application configuration.
+    path_params : dict
+        Parameters used to format the URL. For example, if the value of the
+        configuration parameter is `http://arxiv.org/abs/{arxiv_id}/`,
+        passing `path_params = {'arxiv_id': '1901.00123'}` would generate the
+        URL `http://arxiv.org/abs/1901.00123`.
+    params : dict
+        GET request parameters to add to the URL.
+
+    Returns
+    -------
+    str
+
+    Raises
+    ------
+    :class:`.ConfigurationError`
+        Raised if the configuration parameter for `target` cannot be found.
+
+    Examples
+    --------
+
+    .. code-block:: python
+
+       from arxiv.base.urls import config_url
+    """
+    target = target.upper()
+    # Look for the URL on the config of the current app (this will *not* be
+    # base); fall back to the base config if not found.
+    try:
+        url: str = current_app.config.get(f'ARXIV_{target}_URL')
```

Comment:
Just to check how this works: the URL paths are stored as string templates in the flask app context and these are used by just running format on them with the params dict.

I'm not sure but it looks like this is separated from the flask blueprint and routes mechanisms. Is it seperated from those or some how integrated?

Reply:
An idea for something similar that does use routes would be to define the public routes of other services in base and then redefine then in the actual services:
 ```
In [28]: app = create_web_app()

In [29]: app.url_map
Out[29]: 
Map([<Rule '/test' (HEAD, OPTIONS, GET) -> ui.test_page>,
 <Rule '/404' (HEAD, OPTIONS, GET) -> ui.test_404>,
 <Rule '/403' (HEAD, OPTIONS, GET) -> ui.test_403>,
 <Rule '/401' (HEAD, OPTIONS, GET) -> ui.test_401>,
 <Rule '/405' (HEAD, OPTIONS, GET) -> ui.test_405>,
 <Rule '/413' (HEAD, OPTIONS, GET) -> ui.test_413>,
 <Rule '/400' (HEAD, OPTIONS, GET) -> ui.test_400>,
 <Rule '/500' (HEAD, OPTIONS, GET) -> ui.test_500>,
 <Rule '/static/base/<filename>' (HEAD, OPTIONS, GET) -> base.static>,
 <Rule '/static/<filename>' (HEAD, OPTIONS, GET) -> static>])

In [30]: def pdf(id): print("I'm a placeholder")

In [31]: app.add_url_rule('/pdf/<string:id>', 'pdf',pdf)

In [32]: app.url_map
Out[32]: 
Map([<Rule '/test' (HEAD, OPTIONS, GET) -> ui.test_page>,
 <Rule '/404' (HEAD, OPTIONS, GET) -> ui.test_404>,
 <Rule '/403' (HEAD, OPTIONS, GET) -> ui.test_403>,
 <Rule '/401' (HEAD, OPTIONS, GET) -> ui.test_401>,
 <Rule '/405' (HEAD, OPTIONS, GET) -> ui.test_405>,
 <Rule '/413' (HEAD, OPTIONS, GET) -> ui.test_413>,
 <Rule '/400' (HEAD, OPTIONS, GET) -> ui.test_400>,
 <Rule '/500' (HEAD, OPTIONS, GET) -> ui.test_500>,
 <Rule '/static/base/<filename>' (HEAD, OPTIONS, GET) -> base.static>,
 <Rule '/static/<filename>' (HEAD, OPTIONS, GET) -> static>,
 <Rule '/pdf/<id>' (HEAD, OPTIONS, GET) -> pdf>])```

Reply:
That's an excellent idea. Will take a stab and update the PR

Reply:
Great!